app.py

```diff
@@ -15,9 +15,262 @@
+
+    df_errors = pd.DataFrame()
+    df_errors["result_errors"] = df["result_errors"]
+    df_errors["message"] = df_errors.apply(lambda row: row.result_errors[0]["details"]["message"], axis=1)
+    df_errors["tokenized_message"] = df_errors.apply(lambda row: split_text(row.message), axis=1)
+    df_errors["exit_status"] = df_errors.apply(lambda row: row.tokenized_message[0], axis=1)
+    df_errors["exit_status_ERROR"] = df_errors.apply(lambda row: get_error(row.exit_status), axis=1)
+    df_errors["command"] = df_errors["tokenized_message"].str[1]
+    df_errors["ERROR"] = df_errors["tokenized_message"].str[-2]
+    df_errors["error_info"] = df_errors["tokenized_message"].str[-4]
+    df_errors["error_label"] = df_errors["tokenized_message"].str[-5]
+    df_errors["cwd_info"] = df_errors["tokenized_message"].str[2:-6]
+    df_errors["error_for_analysis"] = df_errors["error_info"] + df_errors["error_label"]
+
+    df_errors["exit_status_ERROR_processed"] = df_errors.apply(
+        lambda row: process_exit_status_error(row.exit_status_ERROR), axis=1
+    )
+    df_errors["error_info_processed"] = df_errors.apply(lambda row: process_error_info(row.error_info), axis=1)
+    df_errors["message_processed"] = df_errors["exit_status_ERROR_processed"] + " " + df_errors["error_info_processed"]
+    df_errors["message_processed"] = df_errors.apply(lambda row: check_is_instance(row.message_processed), axis=1)
+    df_errors["message_processed"] = df_errors.apply(lambda row: remove_stopwords(row.message_processed), axis=1)
+    return df_errors
+
+
+def cluster_errors(*, solver_path: str, max_d: int) -> Tuple[KNeighborsClassifier, TfidfVectorizer]:
+    """Create model and train it on solver errors, model can then classify errors from solver."""
+    df_errors = preprocess_document(solver_path)
+    vectorizer = TfidfVectorizer(stop_words={"english"})
+    x = vectorizer.fit_transform(df_errors["message_processed"])
+    x_arr = x.toarray()
+    z = linkage(x_arr, "ward")
+    clusters = fcluster(z, max_d, criterion="distance")
+    _LOGGER.debug("Clusters: %r", set(clusters))
+
+    def get_data_from_cluster(df, clusters, cluster_number):
+        indices = [i for i, x in enumerate(clusters) if x == cluster_number]
+        df_grouped = df.iloc[indices]
+        return df_grouped
+
+    try:
+        if not os.path.exists("clusters"):
+            os.makedirs("clusters")
+        for i in range(1, len(set(clusters)) + 1):
+            _LOGGER.debug("Cluster %r", i)
+            data = get_data_from_cluster(df_errors, clusters, i)
+            _LOGGER.debug(data)
+            data["message_processed"].to_csv("clusters/{0}.csv".format(i))
+    except OSError as e:
+        _LOGGER.error("Could not create clusters directory: %r", e)
+
+    knn = KNeighborsClassifier()
+
+    knn.fit(x_arr, clusters)
+    return knn, vectorizer
+
+
+def classification(*, model: KNeighborsClassifier, vectorizer: TfidfVectorizer, classify_dataset: str) -> List:
```

Comment:
Is the list returned a list of strings?

```suggestion
def classification(*, model: KNeighborsClassifier, vectorizer: TfidfVectorizer, classify_dataset: str) -> List[str]:
```

Reply:
List of integers